internal_api.go

```diff
@@ -25,6 +25,14 @@ const (
 
 // addPolicy adds a rule to the current policy.
 func (e *Enforcer) addPolicy(sec string, ptype string, rule []string) (bool, error) {
+	if e.adapter != nil && e.autoSave {
```

Comment:
I recommend call model.HasPolicy to check if the policy can be added to adapter.

Reply:
@nodece my only question around this is that this check is already done in the model mutation code. So we'd be duplicating that work. In the case of AddPolicies (or another plural action) we'd be potentially iterating over everything to check again. That isn't great.

We could move the ` model.HasPolicy(...)` check outside of the functions in policy.go?

Reply:
I know it's  duplicate work. If not checked, It will add duplicate policy :(  
`adapter` may have added a unique index on table, but I'm not sure.

Reply:
> We could move the model.HasPolicy(...) check outside of the functions in policy.go?

You can try.

Reply:
I think there is 2 easy options.

1. We move all `model.HasPolicy(...)` checks outside of `policy.go` into the calling code. We would also have to drop the `policy.go` mutation returning a `bool` as it doesn't make sense anymore.

2. Push adapter actions into `policy.go`. This doesn't feel great to me, but it is another option.

Reply:
I have an idea, Let's add a method returns effects policy in policy.go.

Reply:
Not sure I follow @nodece? Do you mean call a func from within `policy.go` that will save to the adapter?

Reply:
Sounds good,  could you try?

Reply:
I will have a look a little bit later today mate :)

Reply:
@speza Looks good. 

cc @hsluoyz

Reply:
@nodece I don't think its particularly pretty! But it allows us to achieve calling the persist adapter without having to pass the whole enforcer or adatper through. Problem is that codeclimate dislikes it! I fear that doing anything nicer at this stage will require a lot more refactoring.

Reply:
It's breaking model struct, we also can add a new method in model, like `AddPolicyWith persist`.

Reply:
I think that would increase code duplication quite a bit.

Reply:
Ok, I think the latest change is nicer @nodece @hsluoyz. This now leaves `policy.go` as the model mutator only. I moved the mutation pre-conditions up into `internal_api.go`. Personally, I think this is nicer as we're not mixing adapter logic in the `policy.go` model methods.